File: ArchitectureModel/src/main/java/com/kailin/architecture_model/util/CheckRootUtil.java
package com.kailin.architecture_model.util;

import java.io.File;
import java.util.concurrent.atomic.AtomicReference;

public final class CheckRootUtil {

    private static final AtomicReference<CheckRootUtil> reference = new AtomicReference<>();

    public static CheckRootUtil getInstance() {
        while (true) {
            CheckRootUtil instance = reference.get();
            if (instance != null)
                return instance;

            instance = new CheckRootUtil();
            if (reference.compareAndSet(null, instance))
                return instance;
        }
    }

    private final String[] rootPaths = new String[]{
            "/sbin/",
            "/system/bin/",
            "/system/xbin/",
            "/data/local/xbin/",
            "/data/local/bin/",
            "/system/sd/xbin/",
            "/system/bin/failsafe/",
            "/data/local/"
    };

    public boolean isRooted() {
        return doesFileExists();
    }

    private boolean doesFileExists() {
        for (String path : rootPaths) {
            File file = new File(String.format("%s/su", path));
            if (file.exists()) {
                return true;
            }
        }
        return false;
    }
}

File: app/src/main/java/com/kailin/utillibrary/CheckRootUtil.kt
package com.kailin.utillibrary

import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.util.concurrent.atomic.AtomicReference

class CheckRootUtil private constructor(){

    private val rootPaths = arrayOf(
        "/sbin/",
        "/system/bin/",
        "/system/xbin/",
        "/data/local/xbin/",
        "/data/local/bin/",
        "/system/sd/xbin/",
        "/system/bin/failsafe/",
        "/data/local/"
    )

    val isRooted: Boolean
        get() = checkRootMethod1() || checkRootMethod2() || checkRootMethod3()

    private fun checkRootMethod1(): Boolean {
        val buildTags = android.os.Build.TAGS
        return buildTags != null && buildTags.contains("test-keys")
    }

    private fun checkRootMethod2(): Boolean {
        for (path in rootPaths) {
            val file = File("$path/su")
            if (file.exists())
                return true
        }
        return false
    }

    private fun checkRootMethod3(): Boolean {
        var process: Process? = null
        return try {
            process = Runtime.getRuntime().exec(arrayOf("/system/xbin/which", "su"))
            BufferedReader(InputStreamReader(process!!.inputStream)).readLine() != null
        } catch (t: Throwable) {
            false
        } finally {
            process?.destroy()
        }
    }

    companion object {

        private val reference = AtomicReference<CheckRootUtil>()

        fun instance(): CheckRootUtil {
            while (true) {
                var checkRootUtil: CheckRootUtil? = reference.get()
                if (checkRootUtil != null)
                    return checkRootUtil

                checkRootUtil = CheckRootUtil()
                if (reference.compareAndSet(null, checkRootUtil))
                    return checkRootUtil
            }
        }
    }
}
